p4-16/spec/P4-16-spec.mdk

```diff
@@ -3620,6 +3635,10 @@ parser Tcp_option_parser(packet_in b, out Tcp_option_stack vec) {
 }
 ~End P4Example
 
+Two header unions can be compared for equality (==) or inequality (!=)
+if they have the exact same type.  The unions are equal if and only if
```

Comment:
Should we use the same phrasing as for structs here? (i.e., "same type and fields can be recursively...") I'm not sure why this is "exact same type."

Reply:
Perhaps this is opening a can of worms by asking this question, but it seems that most places in P4_16 use what I think @jnfoster called "structural typing", which I believe means that two types are considered "compatible" for purposes of comparison, assignment, and passing as a run-time parameter, if they have the same number of sub-elements, in the same relative order, and recursively those also have compatible types.  For example, the existing spec and open source tools treat any values typedef'd with different names, but with the same `bit<17>` underlying type, as compatible.

This is as opposed to "nominal typing", I think, where such an assignment between two typedef'd values with different typedef names, but the same underlying `bit<17>` type, would require an explicit cast or it would be a type error?

If I've got those definitions at least approximately correct, then it sounds like @jnfoster's comment here is that "exact same type" is misleading, if structural typing is allowed for determining whether two values can be compared via `==` or `!=`.  They don't have to be the exact same type, just structurally the same.

Reply:
typedef just introduces another name for the same type, but the type is essentially the same.
This is like C.
Everywhere we require types to be the same, not structurally the same.

Reply:
So is there a reason that one of these proposed additions says "exact same types", but the others do not?